### PR TITLE
Make add command save webpages to disk

### LINKED ISSUE #22: Offline support: Basic implementation
As we want to eventually support offline usage, we should start saving web pages to disk.

Part of #28

### PR DESCRIPTION
**Pull request by thomastanck**
Resolves #22.

Next part after this is to allow loading of local webpages into the WebView.

# Summary of changes

- Added ArticleStorage and related classes to manage saving articles to disk. No articles are kept in memory. Instead, a Path can be obtained in order to load the article required into a WebView.
- Modified UserPrefs and Model accordingly to include the right paths and methods for ArticleStorage.
- Refactored URLs used for Network tests into TypicalEntries.
- Added unit and integration tests.
- Connected article saving to the `add` command.

### CODE REVIEW

File: src/main/java/seedu/address/storage/DataDirectoryArticleStorage.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    }
+
+    /**
+     * Converts the given url to a filename that will be used to write to.
+     */
+    private String urlToFilename(String url) throws NoSuchAlgorithmException {
+        String lowercaseUrl = url.toLowerCase();
+
+        try {
+            // We hash the URL with sha-256, truncate it to 128 bits so it's shorter, then encode it in base32
```
Comment (qjqqyy) [inline src/main/java/seedu/address/storage/DataDirectoryArticleStorage.java:49]: this feels like overkill....

File: src/main/java/seedu/address/network/Network.java [hunk 2]
```diff
@@ -27,4 +27,14 @@ public static InputStream fetchAsStream(String url) throws IOException {
     public static String fetchAsString(String url) throws IOException {
         return new String(new URL(url).openStream().readAllBytes(), "utf-8");
```
Comment (rlrh) [inline src/main/java/seedu/address/network/Network.java]: Shall we refactor this to use `fetchAsBytes`?

**Review by rlrh — APPROVED** [review]
Very nice work!

Just a question: why do we not store the downloaded files with a `.htm` or `.html` extension?

**Comment by thomastanck** [comment]
@rlrh thanks for the review!

I left it out for the moment as that wasn't necessary to meet the requirements and also because I would prefer if the extension on the url matched the extension saved (may not always be .html). It also complicates things when there is no extension.

File: src/main/java/seedu/address/storage/DataDirectoryArticleStorage.java (hunk 1 above)
Comment (thomastanck) [reply src/main/java/seedu/address/storage/DataDirectoryArticleStorage.java:49]: We need some way of shortening the URL to reduce chances of hitting the 260 character path limit on Windows. Using truncated sha-256 is a pretty good and fast way of ensuring that. Finally we also need to encode in base32 in order to store the resultant hash in a way that doesn't conflict with the filesystem.

**Pull request merged**